java/ql/src/experimental/Security/CWE/CWE-094/ScriptInjection.ql

```diff
@@ -14,31 +14,62 @@ import java
 import semmle.code.java.dataflow.FlowSources
 import DataFlow::PathGraph
 
+/** A method of ScriptEngine that allows code injection. */
 class ScriptEngineMethod extends Method {
   ScriptEngineMethod() {
     this.getDeclaringType().getASupertype*().hasQualifiedName("javax.script", "ScriptEngine") and
     this.hasName("eval")
+    or
+    this.getDeclaringType().getASupertype*().hasQualifiedName("javax.script", "Compilable") and
+    this.hasName("compile")
+    or
+    this.getDeclaringType().getASupertype*().hasQualifiedName("javax.script", "ScriptEngineFactory") and
+    this.hasName(["getProgram", "getMethodCallSyntax"])
```

Comment:
For `getProgram` and `getMethodCallSyntax` any argument can all script injection, however in line 76 you are restricting it to `ma.getArgument(0)`.
Might be good to instead add a predicte here, e.g. `int getAnInjectableParamIndex()`, which returns the appropriate index so the code in line 76 can use that.

Reply:
@Marcono1234 I've made the requested change. Please let me know if it's not what you want.

Reply:
Thanks! My suggestion was not well thought through, for `getProgram` and `getMethodCallSyntax` the result index value would still have to be bound, which you have worked around by using `-1`.
Your current code should work fine, but maybe it would be slightly better to change the predicate to:
```ql
bindingset[index]
predicate isInjectableArgIndex(int index) {
  if
    this.getDeclaringType()
        .getASupertype*()
        .hasQualifiedName("javax.script", "ScriptEngineFactory")
  then any()
  else index = 0
}
```
This would avoid the 'magic' value `-1`.

Reply:
Given this is a bit messy and there's only one use site, just inline the method.

Reply:
Thanks @Marcono1234 for the code. I've made the change.

@smowton - sorry I didn't get what you mean. Would you please explain what you meant "inline the method"? Thanks.

Reply:
Consider
```
int f() { return g() + 1; }
int g() { return 1; }
```
To inline g() is to copy its body over the callsite, like:
```
int f() { return 1 + 1; }
// g() no longer needed
```

Reply:
I've updated the implementation. Please let me know if it's not what you want. Thanks.